components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ServiceURL.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.core.model;
+
+import org.apache.commons.collections.MapUtils;
+import org.apache.commons.lang.StringUtils;
+import org.wso2.carbon.identity.core.URLBuilderException;
+import org.wso2.carbon.identity.core.util.IdentityCoreConstants;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.StringJoiner;
+
+/**
+ * Service URL representation.
+ */
+public class ServiceURL {
+
+    private String protocol;
+    private String hostName;
+    private int port;
+    private String urlPath;
+    private Map<String, String> parameters;
+    private String fragment;
+    private String resolvedParamsString;
+    private String absoluteUrl;
+    private String relativeUrl;
+
+    public ServiceURL(String protocol, String hostName, int port, String urlPath, Map<String, String> parameters,
+                      String fragment) throws URLBuilderException {
+
+        this.protocol = protocol;
+        this.hostName = hostName;
+        this.port = port;
+        this.urlPath = urlPath;
+        this.parameters = parameters;
+        this.fragment = fragment;
+        setParamsString();
+        this.absoluteUrl = getAbsoluteUrl();
+        this.relativeUrl = getRelativeUrl();
+    }
+
+    /**
+     * This method is called to get the protocol of {@link ServiceURL}.
+     *
+     * @return String of the protocol.
+     */
+    public String getProtocol() {
+
+        return protocol;
+    }
+
+    /**
+     * This method is called to get the host name of {@link ServiceURL}.
+     *
+     * @return String of the host name.
+     */
+    public String getHostName() {
+
+        if (hostName.endsWith("/")) {
+            hostName = hostName.substring(0, hostName.length() - 1);
+        }
```

Comment:
Should this check be here ?
Ideally it hostname here should not set with '/'. It should be validated at the URL builder level

Reply:
Yes, will move it to the DefaultServiceURLBuilder class.